### PR TITLE
Force EOS at max_tokens with importance-weight correction

### PR DESCRIPTION
**Pull request by ClementeP**
SequenceModel previously truncated sequences without an EOS token when hitting max_tokens, producing samples that did not target the length-conditioned distribution. At the boundary step we now swap the proposal for a point mass on EOS and apply the corresponding IS correction via a new TokenSampler.logw_eos method, so particles are properly weighted with respect to the target conditioned on |y| <= max_tokens; every returned sequence ends with EOS.

SetTokenSampler overrides logw_eos to use complete(ctx) - prefix(ctx), avoiding materializing the full logw_next vector that set samplers are designed to skip during regular sampling.

Tests updated to reflect the new boundary semantics and a focused prefix-overlap test added to cover a case the existing property test was not exercising.

### CODE REVIEW

File: genlm/control/sampler/token.py [hunk 1]
```diff
@@ -36,6 +36,16 @@ async def start_weight(self):
         """Compute the weight of the empty sequence under the target potential."""
         return await self.target.prefix([])
 
+    async def logw_eos(self, context):
+        """
+        Returns teh weight of EOS, used to forse termination in sequence.py.
```
Comment (samuki) [inline genlm/control/sampler/token.py]: Spelling: "teh"-> "the", "forse" -> "force"

File: genlm/control/sampler/sequence.py [hunk 2]
```diff
@@ -289,16 +296,32 @@ async def start(self):
...
+            print(self.__repr__())
 
-        inf_weight = self.weight == float("-inf")
-        if inf_weight:
+        # Advance the context: either sample, or force EOS at the max_tokens
+        # boundary. Forcing EOS is equivalent to swapping the proposal for a
+        # point mass at EOS for that step; its IS correction is therefore the
+        # target's unnormalized log-weight on EOS (see TokenSampler.logw_eos).
+        if self.max_tokens == 1:
+            self.score(await self.unit_sampler.logw_eos(self.token_ctx))
```
Comment (samuki) [inline genlm/control/sampler/sequence.py:307]: I think this could crash in the `MultiTokenUnitSampler`, since it has a nested context. We could overwrite logw_eos in `MultiTokenUnitSampler` to flatten the unit, e.g.,

```
async def logw_eos(self, context):
    return await self.subunit_sampler.logw_eos(flatten_units(context))
```

File: tests/sampler/test_seq_sampler.py [hunk 3]
```diff
@@ -134,11 +134,53 @@ async def test_smc_weights(params):
```
Comment (samuki) [inline tests/sampler/test_seq_sampler.py:134]: Could we also add tests for the other samplers?

**Review by samuki — CHANGES_REQUESTED** [review]

File: genlm/control/sampler/sequence.py (hunk 2 above)
Comment (vicky-xef) [reply genlm/control/sampler/sequence.py:307]: we can include a test to test this!

File: genlm/control/sampler/token.py [hunk 4]
```diff
@@ -36,6 +36,16 @@ async def start_weight(self):
...
+        
+        Subclasses may override with cheaper, sampler-specific computations
+        that avoid materializing the full ``logw_next`` vector.
+        """
+        logws = await self.target.logw_next(context)
+        return logws[self.target.eos]
+
     async def forward(self):
         parent = self.parent  # For some reason, need to hold onto this reference.
```
Comment (vicky-xef) [inline genlm/control/sampler/token.py]: remove comment here

File: tests/sampler/test_seq_sampler.py (hunk 3 above)
Comment (vicky-xef) [reply tests/sampler/test_seq_sampler.py:134]: Yes, I agree! And include edge cases (e.g. max_tokens=1)

**Review by vicky-xef — CHANGES_REQUESTED** [review]

File: genlm/control/sampler/token.py (hunk 1 above)
Comment (ClementeP) [reply genlm/control/sampler/token.py]: Fixed this

File: genlm/control/sampler/sequence.py (hunk 2 above)
Comment (ClementeP) [reply genlm/control/sampler/sequence.py:307]: Good point @samuki , thanks for spotting it. It should be fixed now, let me know what you think.

File: tests/sampler/test_seq_sampler.py (hunk 3 above)
Comment (ClementeP) [reply tests/sampler/test_seq_sampler.py:134]: Good point, now we have tests for all the samplers and the edge cases.

File: genlm/control/sampler/token.py (hunk 4 above)
Comment (ClementeP) [reply genlm/control/sampler/token.py]: Right removed it.

**Review by samuki — APPROVED** [review]

**Review by vicky-xef — APPROVED** [review]
